### PR TITLE
Shell/5560

### PR DESCRIPTION
**Pull request by dsd**
https://github.com/endlessm/eos-shell/issues/5560

### CODE REVIEW

File: drivers/input/keyboard/gpio_keys_ec100.c [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+	kfree(ddata);
+
+	return 0;
+}
+
+static struct platform_driver gpio_keys_device_driver = {
+	.probe		= gpio_keys_probe,
+	.remove		= gpio_keys_remove,
+	.driver		= {
+		.name	= "gpio-keys",
```
Comment (carlocaione) [inline drivers/input/keyboard/gpio_keys_ec100.c]: nit: can we call it "ec100-gpio-keys" to differentiate it from the gpio-keys driver?

File: drivers/input/keyboard/gpio_keys_ec100.c [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+	if (!pdata) {
+		pdata = gpio_keys_get_devtree_pdata(dev);
+		if (IS_ERR(pdata))
+			return PTR_ERR(pdata);
+	}
+
+	ddata = kzalloc(sizeof(struct gpio_keys_drvdata) +
+			pdata->nbuttons * sizeof(struct gpio_button_data),
+			GFP_KERNEL);
+	input = input_allocate_device();
```
Comment (carlocaione) [inline drivers/input/keyboard/gpio_keys_ec100.c]: devm_input_allocate_device ?

File: drivers/input/keyboard/gpio_keys_ec100.c [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+	struct input_dev *input;
+	int i, error;
+
+	if (!pdata) {
+		pdata = gpio_keys_get_devtree_pdata(dev);
+		if (IS_ERR(pdata))
+			return PTR_ERR(pdata);
+	}
+
+	ddata = kzalloc(sizeof(struct gpio_keys_drvdata) +
```
Comment (carlocaione) [inline drivers/input/keyboard/gpio_keys_ec100.c]: devm_kzalloc ?

File: drivers/input/keyboard/gpio_keys_ec100.c [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	struct input_dev *input = ddata->input;
+	int i;
+
+	for (i = 0; i < ddata->pdata->nbuttons; i++)
+		gpio_remove_key(&ddata->data[i]);
+
+	input_unregister_device(input);
+	kfree(ddata->pdata);
+	kfree(ddata);
+
```
Comment (carlocaione) [inline drivers/input/keyboard/gpio_keys_ec100.c]: if we switch to devm_\* we do not need these.

File: drivers/input/keyboard/gpio_keys_ec100.c [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+{
+	struct device *dev = &pdev->dev;
+	int error;
+
+	bdata->input = input;
+	bdata->button = button;
+	input_set_capability(input, EV_KEY, button->code);
+
+	amlogic_gpio_request(button->gpio, "gpio_keys");
+    amlogic_gpio_direction_input(button->gpio, "gpio_keys");
```
Comment (carlocaione) [inline drivers/input/keyboard/gpio_keys_ec100.c]: nit: indentation

File: drivers/input/keyboard/gpio_keys_ec100.c [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+	struct gpio_button_data *bdata = dev_id;
+	const struct gpio_keys_button *button = bdata->button;
+	struct input_dev *input = bdata->input;
+
+	input_event(input, EV_KEY, button->code, 1);
+	input_sync(input);
+
+	input_event(input, EV_KEY, button->code, 0);
+	input_sync(input);
+
```
Comment (carlocaione) [inline drivers/input/keyboard/gpio_keys_ec100.c:55]: Why do we report 1 and then 0?

Comment (magcius) [inline drivers/input/keyboard/gpio_keys_ec100.c:55]: Press and then release.

**Pull request merged**